gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```
@@ -157,32 +164,111 @@ public void handleResponse(WriteObjectResponse response) {
   }
 
   @Override
-  public void startUpload(InputStream pipeSource) {
-    // Given that the two ends of the pipe must operate asynchronous relative
-    // to each other, we need to start the upload operation on a separate thread.
+  public boolean isOpen() {
+    return (writeBuffer != null);
+  }
+
+  @Override
+  public synchronized int write(ByteBuffer buffer) throws IOException {
+    checkState(initialized, "initialize() must be invoked before use.");
+    int bytesWritten = 0;
+    if (!isOpen()) {
+      throw new ClosedChannelException();
+    }
+
+    // No point in writing further if upload failed on another thread.
+    if (uploadOperation.isDone()) {
+      waitForCompletionAndThrowIfUploadFailed();
+    }
+
+    while (buffer.remaining() > 0) {
+      int bytesToWrite = 0;
+      if (lastChunk != null) {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE - lastChunk.size());
+        lastChunk = lastChunk.concat(ByteString.copyFrom(buffer, bytesToWrite));
+      } else {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE);
+        lastChunk = ByteString.copyFrom(buffer, bytesToWrite);
+      }
+      bytesWritten += bytesToWrite;
+      if (lastChunk.size() == MAX_BYTES_PER_MESSAGE) {
+        try {
+          writeBuffer.put(lastChunk);
+          lastChunk = null;
+        } catch (InterruptedException e) {
+          throw new IOException(
+              String.format(
+                  "Failed to write %d bytes in '%s'", buffer.remaining(), getResourceString()),
+              e);
+        }
+      }
+    }
+
+    if (uploadOperation.isDone()) {
+      waitForCompletionAndThrowIfUploadFailed();
+    }
+    return bytesWritten;
+  }
+
+  /** Initialize this channel object for writing. */
+  public void initialize() throws IOException {
```

Should we throw an exception if init is called more than once?